src/puya/awst/nodes.py

```diff
@@ -962,6 +962,29 @@ def accept(self, visitor: ExpressionVisitor[T]) -> T:
         return visitor.visit_tuple_item_expression(self)
 
 
+@attrs.frozen
+class NamedTupleExpression(Expression):
+    """
+    Defines a new named tuple with the specified elements.
+    Elements are evaluated in mapping order.
+    """
+
+    wtype: wtypes.WTuple
+    values: Mapping[str, Expression] = attrs.field(converter=immutabledict)
+
+    @values.validator
+    def _validate_values(self, _instance: object, values: Mapping[str, Expression]) -> None:
+        if values.keys() != self.wtype.fields.keys():
```

Comment:
This should probably first validate that `WTuple.names` is not `None` - if it is then that's an `InternalError`.